lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs

```diff
@@ -108,158 +78,85 @@ internal sealed class UpdateTaskHost<TMapInput, TMapOutput, TResult> : ITask, IO
         /// </summary>
         /// <param name="memento"></param>
         /// <returns></returns>
-        public byte[] Call(byte[] memento)
+        protected override byte[] TaskBody(byte[] memento)
         {
-            Logger.Log(Level.Info, "Entering UpdateTaskHost Call().");
-            var updateResult = _updateTask.Initialize();
-            int iterNo = 0;
+            UpdateResult<TMapInput, TResult> updateResult = null;
             try
             {
-                while (updateResult.HasMapInput && !_cancellationSource.IsCancellationRequested)
-                {
-                    iterNo++;
-
-                    using (
-                        var message = new MapInputWithControlMessage<TMapInput>(updateResult.MapInput,
-                            MapControlMessage.AnotherRound))
-                    {
-                        _dataAndControlMessageSender.Send(message);
-                    }
-
-                    var input = _dataReceiver.Reduce(_cancellationSource);
-
-                    if (_invokeGC)
-                    {
-                        Logger.Log(Level.Verbose, "Calling Garbage Collector");
-                        GC.Collect();
-                        GC.WaitForPendingFinalizers();
-                    }
-
-                    updateResult = _updateTask.Update(input);
-
-                    if (updateResult.HasResult)
-                    {
-                        _resultHandler.HandleResult(updateResult.Result);
-                    }
-                }
-                if (!_cancellationSource.IsCancellationRequested)
-                {
-                    MapInputWithControlMessage<TMapInput> stopMessage =
-                        new MapInputWithControlMessage<TMapInput>(MapControlMessage.Stop);
-                    _dataAndControlMessageSender.Send(stopMessage);
-                }
+                updateResult = _updateTask.Initialize();
             }
-            catch (OperationCanceledException e)
+            catch (Exception e)
             {
-                Logger.Log(Level.Warning,
-                    "Received OperationCanceledException in UpdateTaskHost with message: {0}.",
-                    e.Message);
+                HandleTaskAppException(e);
             }
-            catch (Exception e)
+
+            int iterNo = 0;
+            while (updateResult.HasMapInput && !_cancellationSource.IsCancellationRequested)
```

Comment:
Theoretically it's possible to get to this code path with `updateResult = null` (if `_updateTask.Initialize` threw exception but at the same time another evaluator failed so Driver sent cancel command, and `HandleTaskAppException` didn't throw exception). Let's be paranoid and check for null